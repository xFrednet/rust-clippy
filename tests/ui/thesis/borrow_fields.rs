//@rustc-env: CLIPPY_PRINT_MIR=1

struct A {
    field: String,
}

struct Magic<'a> {
    a: &'a String,
}

const DUCK: u32 = 10;

#[forbid(clippy::borrow_pats)]
fn magic(input: &A, cond: bool) -> &str {
    if cond { "default" } else { &input.field }
}

impl A {
    // #[warn(clippy::borrow_pats)]
    fn borrow_self(&self) -> &A {
        self
    }

    // #[warn(clippy::borrow_pats)]
    fn borrow_field_direct(&self) -> &String {
        &self.field
    }

    // #[warn(clippy::borrow_pats)]
    fn borrow_field_deref(&self) -> &str {
        &self.field
    }

    fn borrow_field_or_default(&self) -> &str {
        if self.field.is_empty() {
            "Here be defaults"
        } else {
            &self.field
        }
    }

    fn borrow_field_into_mut_arg<'a>(&'a self, magic: &mut Magic<'a>) {
        magic.a = &self.field;
    }
}

fn main() {}
