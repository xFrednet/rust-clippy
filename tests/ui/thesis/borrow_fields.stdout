## MIR:
bb0:
    StorageLive(_3)
    _3 = _2
    switchInt(move _3) -> [0: bb2, otherwise: bb1]

bb1:
    StorageLive(_4)
    _4 = const "default"
    _0 = &(*_4)
    StorageDead(_4)
    goto -> bb4

bb2:
    StorageLive(_5)
    StorageLive(_6)
    StorageLive(_7)
    _7 = &((*_1).0: std::string::String)
    _6 = &(*_7)
    _5 = <std::string::String as std::ops::Deref>::deref(move _6) -> [return: bb3, unwind: bb5]

bb3:
    _0 = &(*_5)
    StorageDead(_6)
    StorageDead(_7)
    StorageDead(_5)
    goto -> bb4

bb4:
    StorageDead(_3)
    return

bb5:
    resume

